README.md
# PageView Widget Indicator

This is a widget that syncronizes with your pageView and display a nice indicator on a widget.

<img src="https://github.com/andersonfds/pageview_widget_indicator/blob/main/assets/pageview_widget_example.gif?raw=true" alt="page view indicator example" width="200" />

<img src="https://github.com/andersonfds/pageview_widget_indicator/blob/main/assets/pageview_dots_example.gif?raw=true" alt="page view dots example" width="200" />

Usage:

[example/pageview_widget_example](https://github.com/andersonfds/pageview_widget_indicator/blob/db5a51c3cb5d4e7f8ca7f5a6d4b8a111d8ff9393/example/pageview_widget_example/pageview_widget_example.dart#L62-L94)

```dart
@override
Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
        title: Text(widget.title),
        bottom: PreferredSize(
            preferredSize: Size.fromHeight(kToolbarHeight),
            child: PageViewIndicator(
            pageController: controller,
            itemWidth: 50.0,
            itemCount: 10,
            builder: (context, index, isSelected) => Padding(
                padding: const EdgeInsets.symmetric(vertical: 4.0),
                child: Material(
                color: Colors.white.withOpacity(isSelected ? 0.25 : 0.0),
                borderRadius: BorderRadius.circular(12.0),
                clipBehavior: Clip.antiAlias,
                child: InkWell(
                    onTap: () {
                    controller.jumpToPage(index);
                    },
                    child: Center(
                    child: Text(
                        '${index + 1}',
                        textAlign: TextAlign.center,
                    ),
                    ),
                ),
                ),
            ),
            ),
        ),
        ),
        body: PageView.builder(
        controller: controller,
        itemCount: 10,
        itemBuilder: (context, index) => Container(
            color: Colors.red[(index + 1) % 10 * 100],
            child: Center(
            child: Text(
                'shade ' + ((index + 1) % 10 * 100).toString(),
                style:
                    CupertinoTheme.of(context).textTheme.navLargeTitleTextStyle,
            ),
            ),
        ),
        ),
    );
}
```

To see all examples, head to [this link](https://github.com/andersonfds/pageview_widget_indicator/tree/main/example/lib).

**Sidenotes:**

This project is licensed under MIT.
Feel free to open issues or make pull requests.


CHANGELOG.md
# Changelog

## 0.0.4

* Migrated to null-safety

## 0.0.3+3

* Fixed some performance issues
* Added padding support

## 0.0.3+2

* Updated documentation
* Included more usage examples

## 0.0.3+1

* Updated documentation

## 0.0.3

* Updated documentation
* Removed nullsafety to reach everyone

## 0.0.2

* Fixed licensing

## 0.0.1

* Initial Release


lib/pageview_widget_indicator.dart
library pageview_widget_indicator;

import 'package:flutter/material.dart';
import 'package:scrollable_positioned_list/scrollable_positioned_list.dart';

/// The alignment of the PageView items
///
/// When scrolling the widget it will stick to the chosen position
enum PageViewAlignment {
  start,
  center,
  end,
}

class PageViewIndicator extends StatefulWidget {
  /// The builder function to the elements
  /// [context] is the context of the widget
  /// [index] is the index of the current element
  /// [isSelected] is true if the current element is selected
  final Widget Function(
    BuildContext context,
    int index,
    bool isSelected,
  ) builder;

  /// The page controller
  final PageController pageController;

  /// The Alignment to the scroll
  final PageViewAlignment alignment;

  /// The width of the item
  final double itemWidth;

  /// The height of the item
  final double height;

  /// The itemCount for the builder property
  final int itemCount;

  /// The initial item that should be selected
  /// if not specified it will gett from the [pageController]
  final int? initialPage;

  /// Inner padding of the lit
  final EdgeInsetsGeometry? padding;

  const PageViewIndicator({
    Key? key,
    required this.builder,
    required this.pageController,
    required this.itemWidth,
    required this.itemCount,
    this.initialPage,
    this.height = kToolbarHeight,
    this.alignment = PageViewAlignment.center,
    this.padding,
  }) : super(key: key);

  @override
  _PageViewIndicatorState createState() => _PageViewIndicatorState();
}

class _PageViewIndicatorState extends State<PageViewIndicator> {
  final ItemScrollController _scrollController = ItemScrollController();
  int _page = 0;

  Future<void> setPage(int value) async {
    if (_page != value) {
      await _scrollController.scrollTo(
        index: value,
        duration: Duration(milliseconds: 300),
        curve: Curves.easeInOut,
        alignment: alignment,
      );
      _page = value;
    }
  }

  double get alignment {
    switch (widget.alignment) {
      case PageViewAlignment.start:
        return 0.0;
      case PageViewAlignment.end:
        return 1.0;
      default:
        // Calculating the percentage we should use considering the width of the widget
        final screenWidth = MediaQuery.of(context).size.width;
        final startOffset = (screenWidth - widget.itemWidth) / 2;
        final inPercentage = startOffset * 100 / screenWidth / 100;
        return inPercentage;
    }
  }

  _onPageControllerChanged() async {
    final newPage = widget.pageController.page?.round();
    if (newPage != null && newPage != _page) {
      await setPage(newPage);
      setState(() {});
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    widget.pageController.removeListener(_onPageControllerChanged);
    widget.pageController.addListener(_onPageControllerChanged);
    _page = widget.initialPage ?? widget.pageController.initialPage;
    setState(() {});
  }

  @override
  void dispose() {
    super.dispose();
    widget.pageController.removeListener(_onPageControllerChanged);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      height: widget.height,
      child: ScrollablePositionedList.builder(
        physics: ClampingScrollPhysics(),
        padding: widget.padding as EdgeInsets?,
        itemScrollController: _scrollController,
        itemCount: widget.itemCount,
        initialScrollIndex: widget.pageController.initialPage,
        initialAlignment: alignment,
        scrollDirection: Axis.horizontal,
        itemBuilder: (context, index) => SizedBox(
          width: widget.itemWidth,
          child: widget.builder(
            context,
            index,
            index == _page,
          ),
        ),
      ),
    );
  }
}


example/lib/pageview_dots_example.dart
import 'package:flutter/material.dart';
import 'package:pageview_widget_indicator/pageview_widget_indicator.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final PageController controller = PageController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('example')),
      body: SingleChildScrollView(
        child: Stack(
          children: [
            SizedBox(
              height: 300,
              child: PageView.builder(
                controller: controller,
                itemBuilder: (context, index) => Container(
                  color: Colors.red[(index + 1) % 10 * 100],
                  child: Image.network(
                    'https://picsum.photos/seed/$index/200',
                    fit: BoxFit.cover,
                  ),
                ),
              ),
            ),
            Positioned(
              bottom: 0,
              left: 0,
              right: 0,
              child: Center(
                child: SizedBox(
                  width: 100,
                  height: 20,
                  child: PageViewIndicator(
                    height: 30,
                    builder: (context, index, isSelected) => Container(
                      margin: const EdgeInsets.symmetric(horizontal: 2),
                      decoration: BoxDecoration(
                        color: Colors.white.withOpacity(isSelected ? 1.0 : 0.2),
                        shape: BoxShape.circle,
                      ),
                    ),
                    pageController: controller,
                    itemWidth: 15,
                    itemCount: 10,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/pageview_widget_example.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:pageview_widget_indicator/pageview_widget_indicator.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late PageController controller;

  @override
  void initState() {
    super.initState();
    controller = PageController(
      initialPage: 4,
      keepPage: true,
    );
  }

  Widget _myWidget(index, isSelected) => Padding(
        padding: const EdgeInsets.symmetric(vertical: 4.0),
        child: Material(
          color: Colors.white.withOpacity(isSelected ? 0.25 : 0.0),
          borderRadius: BorderRadius.circular(12.0),
          clipBehavior: Clip.antiAlias,
          child: InkWell(
            onTap: () {
              controller.jumpToPage(index);
            },
            child: Center(
              child: Text(
                '${index + 1}',
                textAlign: TextAlign.center,
              ),
            ),
          ),
        ),
      );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        bottom: PreferredSize(
          preferredSize: Size.fromHeight(kToolbarHeight),
          child: PageViewIndicator(
            pageController: controller,
            itemWidth: 50.0,
            itemCount: 10,
            builder: (context, index, isSelected) =>
                _myWidget(index, isSelected),
          ),
        ),
      ),
      body: PageView.builder(
        controller: controller,
        itemCount: 10,
        itemBuilder: (context, index) => Container(
          color: Colors.red[(index + 1) % 10 * 100],
          child: Center(
            child: Text(
              'shade ' + ((index + 1) % 10 * 100).toString(),
              style:
                  CupertinoTheme.of(context).textTheme.navLargeTitleTextStyle,
            ),
          ),
        ),
      ),
    );
  }
}
